prow/git/git.go

```diff
@@ -125,6 +129,14 @@ func (c *Client) unlockRepo(repo string) {
 	c.repoLocks[repo].Unlock()
 }
 
+func defaultExecWithDir(dir, name string, arg ...string) ([]byte, error) {
```

Comment:
I'm not sure this code belongs here, this seems like a hacky abstraction. If we want to abstract the way we exec commnd we should clearly surface it in an interface IMO

Reply:
I agree, using a global variable and overwriting it in the tests is not as clean as breaking out some of the functionality in the client into an interface and allowing it to be injectable.

Reply:
This would definitely be nice as we try to use the:
```golang
for _, tc := range cases {
  t.Run(tc.name, func (t *testing.T) {
    //whatever
  })
}
```
pattern everywhere for more concurrency and speed

Reply:
ok I can do injection in the Client.
strictly speaking t.Run is not enough for more concurrency and speed. By default subtests are running serially. t.Parallel() would be required, and I see it only once in the whole repo:
https://github.com/kubernetes/test-infra/blob/03fd014d7cd6c2f4ed0c9526f5ec453ef2c6b293/prow/apis/prowjobs/v1/types_test.go#L186

Reply:
ok I took a shot at injecting it into the Client. The injected executor should be also used in Repo. it also needed in retryCmd so retryCmd get additional parameter or turns into a method on the executor object. It turns essentially into https://github.com/kubernetes/test-infra/pull/11248 and I don't like it. I don't like creating like 70 lines of boilerplate code that is still very hard to test (how would anyone test runWithRetries there? https://github.com/kubernetes/test-infra/pull/11248/files#diff-ba863e91489bc04d2a9df5e39ea62d96R83
several additional objects created and passed around just for the sake of running Client tests in parallel. Seems quite unreasonable to me.

Reply:
Closing over behavior with interfaces on functional barriers and creating larger objects from compositions of functional interfaces is really a cornerstone of Go. We do it in a lot of other places in the code and it makes for a well-tested system. Not sure what I can say other than it's often best practice. If we had Python we'd do this differently.

Reply:
This is unresolved?

Reply:
This remains unresolved. We are almost 100% consistent in _not_ using this pattern in the rest of the codebase, with one caveat being `time.Now`. I do not think we can make an exception for this PR.

Reply:
@stevekuznetsov @fejta I explained why I disagree with your suggestion earlier in this thread. It makes the code less readable and creates components that are still hard to test.
I am not going to create "enterprise java" style code in go. If it is unacceptable - feel free to close this PR.